lib/middleware/operation.js

```diff
@@ -32,55 +36,92 @@ function findPropertyOperations ({ types, method, term, resource }) {
+    let moreOperations
+    if ('property' in resource) {
+      moreOperations = findPropertyOperations({ resource, method, api }).toArray()
+    } else {
+      const types = clownface({ ...api, term: [...resource.types] })
+      moreOperations = findClassOperations(types, method).toArray()
     }
 
-    const method = req.method === 'HEAD' ? 'GET' : req.method
-    const types = clownface({ ...api, term: [...req.hydra.resource.types] })
-
-    let operations
-    if (req.hydra.term.equals(req.hydra.resource.term)) {
-      // only look for direct operation when the root resource is requested
-      operations = findClassOperations(types, method)
-    }
-    if (!operations) {
-      // otherwise try finding the operation by property usage
-      operations = findPropertyOperations({ types, method, term: req.hydra.term, resource: req.hydra.resource })
+    if (moreOperations.length) {
+      return moreOperations.map(operation => ({ resource, operation }))
     }
 
-    const [operation, ...rest] = (operations || []).toArray()
+    return [{ resource, operation: null }]
+  })
+}
+
+async function invokeOperation (req, res, next) {
+  const { api } = req.hydra
+
+  if (!req.hydra.operation) {
+    const operationMap = mapOperations({ api, res })
+    const allowedMethods = new Set(operationMap
+      .filter(({ operation }) => operation)
+      .flatMap(({ operation }) => operation.out(ns.hydra.method).values))
 
-    if (!operation) {
+    if (!allowedMethods.size) {
       warn('no operations found')
-      let allowedOperations
-      if (req.hydra.term.equals(req.hydra.resource.term)) {
-        allowedOperations = findClassOperations(types)
-      } else {
-        allowedOperations = findPropertyOperations({ types, term: req.hydra.term, resource: req.hydra.resource })
+      if (operationMap.every(({ resource }) => 'property' in resource)) {
+        return next(new createError.NotFound())
       }
-
-      const allowedMethods = new Set(allowedOperations.out(ns.hydra.method).values)
-      res.setHeader('Allow', [...allowedMethods])
       return next(new createError.MethodNotAllowed())
     }
 
-    if (rest.length > 0) {
-      error('Multiple operations found')
-      return next(new Error('Ambiguous operation'))
+    res.setHeader('Allow', [...allowedMethods])
+    return next(new createError.MethodNotAllowed())
+  }
+
+  const handler = await api.loaderRegistry.load(req.hydra.operation.out(code.implementedBy), { basePath: api.codePath })
+
+  if (!handler) {
+    return next(new Error('Failed to load operation'))
+  }
+
+  handler(req, res, next)
+}
+
+function findOperation (operationMiddleware) {
+  return async function (req, res, next) {
+    const method = req.method === 'HEAD' ? 'GET' : req.method
+
+    let operations = mapOperations({ api: req.hydra.api, res, method })
+      .filter(({ operation }) => operation)
+
+    if (operationMiddleware) {
+      operations = operationMiddleware(operations, { req, res })
```

Comment:
Calling the operations middleware hook every time. This way an API might get a chance to select an operation even if none were found otherwise. Or replace those found with something else completely.